File: Semestre_2/Vision_Computadora/Entregas/Python/T5_4_Filtrado_Ruido.py
import matplotlib.pyplot as plt
import numpy as np
from skimage.io import imread
from skimage.color import rgb2gray
from tkinter import Tk, filedialog

file_path = filedialog.askopenfilename(
    title='File Selector',
    filetypes=[("Images", "*.*;*.jpg;*.png;*.jpeg;*.bmp;*.gif")]
)

A = imread(file_path).astype(float) / 255.0
if A.ndim == 3:
    A = rgb2gray(A)

img = A

f = np.fft.fft2(img)
fshift = np.fft.fftshift(f)
magnitude_spectrum = 20 * np.log(np.abs(fshift) + 1)
phase_spectrum = np.angle(fshift)

finv = np.fft.ifft2(f)
f_back = np.abs(finv)

rows, cols = img.shape
mask = np.ones((rows, cols))
band_width = 10
center_col = int(round(cols / 2))

mask[:, center_col-50:center_col-50+band_width] = 0
mask[:, center_col+50-band_width+1:center_col+50+1] = 0

fshift_masked = fshift * mask
magnitude_masked = 20 * np.log(np.abs(fshift_masked) + 1)

f_ishift = np.fft.ifftshift(fshift_masked)
img_restored = np.abs(np.fft.ifft2(f_ishift))

plt.figure()
plt.subplot(2,3,1); plt.imshow(img, cmap='gray');                plt.title('Imagen de entrada');        plt.axis('off')
plt.subplot(2,3,2); plt.imshow(magnitude_spectrum, cmap='gray'); plt.title('Espectro de magnitud');    plt.axis('off')
plt.subplot(2,3,3); plt.imshow(mask, cmap='gray');               plt.title('Mascara de bandas de ruido');plt.axis('off')
plt.subplot(2,3,4); plt.imshow(magnitude_masked, cmap='gray');   plt.title('Masked Spectrum');         plt.axis('off')
plt.subplot(2,3,5); plt.imshow(img_restored, cmap='gray');       plt.title('Imagen restaurada');       plt.axis('off')
plt.subplot(2,3,6); plt.imshow(phase_spectrum, cmap='gray');     plt.title('Espectro de fase');        plt.axis('off')
plt.tight_layout()
plt.show()

File: Semestre_2/Vision_Computadora/Entregas/Matlab/T5_4_Filtrado_Ruido.m
clear;
close all;
clc;

[file,path] = uigetfile({'*.*';'*.jpg';'*.png';'*.jpeg';'*.bmp';'*.gif'}, 'File Selector');
ruta = strcat(path,file);
A = imread(ruta);
img = im2double(A);

% FFT y espectro
f = fft2(img);
fshift = fftshift(f);
magnitude_spectrum = 20*log(abs(fshift)+1);
phase_spectrum = angle(fshift);

% Reconstrucción directa con IFFT
finv = ifft2(f);
f_back = abs(finv);

% Crear máscara para eliminar bandas de ruido
[rows, cols] = size(img);
mask = ones(rows, cols);
band_width = 10;
center_col = round(cols/2);

% Puedes ajustar la posición y ancho de la banda
mask(:, center_col-50:center_col-50+band_width-1) = 0;
mask(:, center_col+50-band_width+1:center_col+50) = 0;

% Aplicar máscara en el dominio frecuencial
fshift_masked = fshift .* mask;
magnitude_masked = 20*log(abs(fshift_masked)+1);

% IFFT de espectro retocado
f_ishift = ifftshift(fshift_masked);
img_restored = abs(ifft2(f_ishift));

% Visualización resultados
figure();
subplot(2,3,1); imshow(img,[]); title('Imagen de entrada'); axis off;
subplot(2,3,2); imshow(magnitude_spectrum,[]); title('Espectro de magnitud'); axis off;
subplot(2,3,3); imshow(mask,[]); title('Mascara de bandas de ruido'); axis off;
subplot(2,3,4); imshow(magnitude_masked,[]); title('Masked Spectrum'); axis off;
subplot(2,3,5); imshow(img_restored,[]); title('Imagen restaurada'); axis off;
subplot(2,3,6); imshow(phase_spectrum,[]); title('Espectro de fase'); axis off;